packages/pg-pool/index.js

```diff
@@ -116,11 +116,22 @@ class Pool extends EventEmitter {
     const pendingItem = this._pendingQueue.shift()
     if (this._idle.length) {
       const idleItem = this._idle.pop()
-      clearTimeout(idleItem.timeoutId)
       const client = idleItem.client
-      const idleListener = idleItem.idleListener
+      clearTimeout(idleItem.timeoutId)
+
+      // TODO(bmc): we need a better state represetation on the client itself
+      // to indicate if it's connecting, idle, running a query, closed, etc...
+      // as right now this hacky fix will only work for the JS client
+      // since the native client doesn't _have_ a connection property
 
-      return this._acquireClient(client, pendingItem, idleListener, false)
+      // remove this client, it's died in the background - this can happen
+      // in aws lambdas - they go idle and the streams are closed without an error
+      if (client.connection && client.connection.stream.readyState !== 'open') {
+        this._remove(client)
```

Comment:
In the current logic: if it notices that the next-available `client` is already dead, then it removes it from the pool.  Afterwards, it instantiates a new client to replace it.

Would there be any advantage to looping on the `client` eviction step?  In pseudo-code:

```
while(this._idle.length) {
  const client = this._idle.pop();

  // ... Other important details I'm ignoring for the moment...

  if(isDead(client)) {
    this.remove(client);
  } else {
    return client
  }
}

if(!this._isFull()) {
  return this.newClient(...);
}

throw new Error('unexpected condition');
```

This seems like it helps in 2 ways:

1. It clears out the dead clients more quickly, and
2. It avoids the need to wait for `this.newClient(..)` when there is already another client available.

Reply:
probably, yeah - good suggestions!

Reply:
hi guys, I want to test this branch, but npm says its missing the version... tried to fork it, put the version (https://github.com/592da/node-postgres) and then compiling failed : `Cannot find module 'pg' `

@brianc

Reply:
Hmm... I suspect that's because this is a mono-repo that is built via [Lerna](https://github.com/lerna/lerna) .  You might need to build it locally first and then link to the build artifacts.  (@brianc might know another approach.  I haven't actually worked with this codebase before).

I don't remember the exact build steps for Lerna off the top of my head, but I think it's something like:

1. Install Lerna
2. Navigate to the root of the repo
3. Run the command:  `lerna bootstrap`

Reply:
> hi guys, I want to test this branch, but npm says its missing the version... tried to fork it, put the version (https://github.com/592da/node-postgres) and then compiling failed : `Cannot find module 'pg' `

if you wanna test it it's probably easiest to just apply the diff of the 1 file i changed inside your `node_modules/pg-pool` folder to `index.js` (or just copy/paste the changes).

Reply:
thanks for your reply @briandamaged, unfortunately, it didn't work, maybe because its also imported from `typeorm`

Reply:
@brianc , having the same issue...
```
2020-02-26T21:38:37.302Z	da5bbf66-06b7-4c69-9c85-ab84b1b777a1	INFO	Error: Connection terminated unexpectedly
    at Connection.<anonymous> (/var/task/src/rest/webpack:/node_modules/pg/lib/client.js:255:1)
    at Object.onceWrapper (events.js:312:28)
    at Connection.emit (events.js:223:5)
    at Connection.EventEmitter.emit (domain.js:475:20)
    at Socket.<anonymous> (/var/task/src/rest/webpack:/node_modules/pg/lib/connection.js:139:1)
    at Socket.emit (events.js:228:7)
    at Socket.EventEmitter.emit (domain.js:475:20)
    at endReadableNT (_stream_readable.js:1185:12)
    at processTicksAndRejections (internal/process/task_queues.js:81:21)```